internal/dcache/rpc/server/handler.go

```diff
@@ -47,43 +57,894 @@ var _ service.ChunkService = &ChunkServiceHandler{}
+		common.Assert(false, "received nil Hello request")
+		return nil, rpc.NewResponseError(rpc.InvalidRequest, "received nil Hello request")
+	}
+
+	log.Debug("ChunkServiceHandler::Hello: Received Hello request: %+v", *req)
+
+	// // TODO:: integration: call cluster manager to get the node ID of this node
+	myNodeID := getNodeUUID()
+	common.Assert(req.ReceiverNodeID == myNodeID, "Received Hello RPC destined for another node", req.ReceiverNodeID, myNodeID)
+
+	// get all the RVs exported by this node
+	myRvList := make([]string, 0)
+	for _, info := range h.rvIDMap {
+		myRvList = append(myRvList, info.rvName)
+	}
+
+	return &models.HelloResponse{
+		ReceiverNodeID: myNodeID,
+		Time:           time.Now().UnixMicro(),
+		RV:             myRvList,
+		MV:             req.MV, // TODO:: discuss: how to fetch the MV list receiver shares with the sender
+	}, nil
 }
 
 func (h *ChunkServiceHandler) GetChunk(ctx context.Context, req *models.GetChunkRequest) (*models.GetChunkResponse, error) {
-	return nil, nil
+	if req == nil || req.Address == nil {
+		log.Err("ChunkServiceHandler::GetChunk: Received nil GetChunk request")
+		common.Assert(false, "received nil GetChunk request")
+		return nil, rpc.NewResponseError(rpc.InvalidRequest, "received nil GetChunk request")
+	}
+
+	log.Debug("ChunkServiceHandler::GetChunk: Received GetChunk request: %+v", *req)
+
+	// check if the chunk address is valid
+	err := h.checkValidChunkAddress(req.Address)
+	if err != nil {
+		log.Err("ChunkServiceHandler::GetChunk: Invalid chunk address [%s]", err.Error())
+		return nil, err
+	}
+
+	rvInfo := h.rvIDMap[req.Address.RvID]
+	mvInfo, _ := rvInfo.getMVInfo(req.Address.MvName)
+
+	// check if the sync has started. If yes, block the chunk operations till the sync is completed
+	mvInfo.blockChunkOps()
+
+	// increment the chunk operation count for this MV
+	mvInfo.incrementChunkOps()
+
+	// decrement the chunk operation count for this MV when the function returns
+	defer mvInfo.decrementChunkOps()
+
+	startTime := time.Now()
+
+	// check if the chunk file is being updated in parallel by some other thread
+	chunkAddress := getChunkAddress(req.Address.FileID, req.Address.RvID, req.Address.MvName, req.Address.OffsetInMB)
+	flock := h.locks.Get(chunkAddress)
+	flock.Lock()
+	defer flock.Unlock()
+
+	cacheDir := rvInfo.cacheDir
+	chunkPath, hashPath := getChunkAndHashPath(cacheDir, req.Address.MvName, req.Address.FileID, req.Address.OffsetInMB)
+	log.Debug("ChunkServiceHandler::GetChunk: chunk path %s, hash path %s", chunkPath, hashPath)
+
+	fh, err := os.Open(chunkPath)
+	if err != nil {
+		log.Err("ChunkServiceHandler::GetChunk: Failed to open chunk file %s [%v]", chunkPath, err.Error())
+		return nil, rpc.NewResponseError(rpc.InternalServerError, fmt.Sprintf("failed to open chunk file %s [%v]", chunkPath, err.Error()))
+	}
+	defer fh.Close()
+
+	// TODO:: integration: call cluster manager to get chunk size
+	chunkSize := getChunkSize()
```

Comment:
note that technically chunk size for different files can be different, so the only way to know the size is to do a stat() call